src/main/java/org/opensearch/security/dlic/rest/api/ConfigUpgradeApiAction.java

```diff
@@ -0,0 +1,395 @@
+    }
+
+    @Override
+    protected EndpointValidator createEndpointValidator() {
+        return new EndpointValidator() {
+
+            @Override
+            public Endpoint endpoint() {
+                return endpoint;
+            }
+
+            @Override
+            public RestApiAdminPrivilegesEvaluator restApiAdminPrivilegesEvaluator() {
+                return securityApiDependencies.restApiAdminPrivilegesEvaluator();
+            }
+
+            @Override
+            public ValidationResult<SecurityConfiguration> entityReserved(SecurityConfiguration securityConfiguration) {
+                // Allow modification of reserved entities
+                return ValidationResult.success(securityConfiguration);
+            }
+
+            @Override
+            public ValidationResult<SecurityConfiguration> entityHidden(SecurityConfiguration securityConfiguration) {
+                // Allow modification of hidden entities
+                return ValidationResult.success(securityConfiguration);
+            }
+
+            @Override
+            public RequestContentValidator createRequestContentValidator(final Object... params) {
+                return new ConfigUpgradeContentValidator(new RequestContentValidator.ValidationContext() {
+                    @Override
+                    public Object[] params() {
+                        return params;
+                    }
+
+                    @Override
+                    public Settings settings() {
+                        return securityApiDependencies.settings();
+                    }
+
+                    @Override
+                    public Map<String, DataType> allowedKeys() {
+                        return Map.of(REQUEST_PARAM_CONFIGS_KEY, DataType.ARRAY);
+                    }
+                });
+            }
+        };
+    }
+
+    /** More permissions validation that default ContentValidator  */
+    static class ConfigUpgradeContentValidator extends RequestContentValidator {
+
+        protected ConfigUpgradeContentValidator(final ValidationContext validationContext) {
+            super(validationContext);
+        }
+
+        @Override
+        public ValidationResult<JsonNode> validate(final RestRequest request, final JsonNode jsonContent) throws IOException {
+            return validateContentSize(jsonContent);
+        }
+    }
+
+    /** Tranforms config changes from a raw PATCH into simplier view */
+    static class ConfigItemChanges {
+
+        private final CType config;
+        private final Map<String, List<String>> itemsGroupedByOperation;
+
+        public ConfigItemChanges(final CType config, final JsonNode differences) {
+            this.config = config;
+            this.itemsGroupedByOperation = classifyChanges(differences);
+        }
+
+        public boolean hasChanges() {
+            return !itemsGroupedByOperation.isEmpty();
+        }
+
+        public void addToNode(final ObjectNode node) {
```

Comment:
could you add a brief comment here with an example to visualize what the output would look like for easier readability?
something like this for roles:
```
{
  "roles": {
    "modify": ["role1", "role2"],
    "delete": ["role3"]
  }
}
```